pkg/controller.v1/tensorflow/status_test.go

```diff
@@ -0,0 +1,609 @@
+					tfJob:                   testutil.NewTFJobWithChief(4, 2),
+					expectedFailedPS:        0,
+					expectedSucceededPS:     0,
+					expectedActivePS:        2,
+					expectedFailedWorker:    4,
+					expectedSucceededWorker: 0,
+					expectedActiveWorker:    0,
+					expectedFailedChief:     0,
+					expectedSucceededChief:  1,
+					expectedActiveChief:     0,
+					restart:                 false,
+					worker0Completed:        false,
+					expectedType:            commonv1.JobSucceeded,
+				},
+				{
+					description:             "Chief is failed and restarting",
+					tfJob:                   testutil.NewTFJobWithChief(4, 2),
+					expectedFailedPS:        0,
+					expectedSucceededPS:     0,
+					expectedActivePS:        2,
+					expectedFailedWorker:    4,
+					expectedSucceededWorker: 0,
+					expectedActiveWorker:    0,
+					expectedFailedChief:     1,
+					expectedSucceededChief:  0,
+					expectedActiveChief:     0,
+					restart:                 true,
+					worker0Completed:        false,
+					expectedType:            commonv1.JobRestarting,
+				},
+			}
+
+			jobNameTemplate := "test-status-%d"
+			for i, c := range testCases {
+				reconciler.Log.Info("testing case", "description", c.description)
+				c.tfJob.SetName(fmt.Sprintf(jobNameTemplate, i))
+				c.tfJob.SetUID(uuid.NewUUID())
+
+				initializeReplicaStatuses(&c.tfJob.Status, tfv1.TFReplicaTypeWorker)
+				initializeReplicaStatuses(&c.tfJob.Status, tfv1.TFReplicaTypeChief)
+				initializeReplicaStatuses(&c.tfJob.Status, tfv1.TFReplicaTypePS)
+
+				setStatusForTest(c.tfJob, tfv1.TFReplicaTypePS, c.expectedFailedPS, c.expectedSucceededPS, c.expectedActivePS, c.restart, c.worker0Completed, testK8sClient)
+				setStatusForTest(c.tfJob, tfv1.TFReplicaTypeWorker, c.expectedFailedWorker, c.expectedSucceededWorker, c.expectedActiveWorker, c.restart, c.worker0Completed, testK8sClient)
+				setStatusForTest(c.tfJob, tfv1.TFReplicaTypeChief, c.expectedFailedChief, c.expectedSucceededChief, c.expectedActiveChief, c.restart, c.worker0Completed, testK8sClient)
+
+				// Adding this section to make sure all pods are created and cached
+				Eventually(func() error {
+					podList := &corev1.PodList{}
+					basicLabels := reconciler.GenLabels(c.tfJob.GetName())
+					selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+						MatchLabels: basicLabels,
+					})
+					if err != nil {
+						return err
+					}
+					listOpt := client.MatchingLabelsSelector{
+						Selector: selector,
+					}
+					err = testK8sClient.List(context.Background(), podList, listOpt)
+					if err != nil {
+						return nil
+					}
+					totalExpectedPodCount := c.expectedFailedPS + c.expectedSucceededPS + c.expectedActivePS +
+						c.expectedFailedWorker + c.expectedSucceededWorker + c.expectedActiveWorker +
+						c.expectedFailedChief + c.expectedSucceededChief + c.expectedActiveChief
+					if len(podList.Items) != int(totalExpectedPodCount) {
```

Comment:
How can we check if different roles's expectation if we sum all of them ?

Reply:
This step is not checking if Pods for each replicaType are created, it's checking all Pods on this test cases are created and are enqueued in the cache of the client. Sometimes test cases failed because Pods are not enqueued in cache.